File: app/src/main/java/com/mmutert/freshfreezer/viewmodel/ItemListViewModel.java
package com.mmutert.freshfreezer.viewmodel;

import android.app.Application;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.work.WorkManager;

import com.mmutert.freshfreezer.data.FrozenItem;
import com.mmutert.freshfreezer.data.ItemNotification;
import com.mmutert.freshfreezer.data.ItemRepository;
import com.mmutert.freshfreezer.util.SortingOption;

import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;


public class ItemListViewModel extends AndroidViewModel {

    public static final String TAG = "FrozenItemViewModel";

    private ItemRepository mItemRepository;

    private LiveData<List<FrozenItem>> mFrozenItems;

    private SortingOption.SortingOrder sortingOrder = SortingOption.SortingOrder.ASCENDING;
    private SortingOption sortingOption = SortingOption.DATE_BEST_BEFORE;


    public ItemListViewModel(@NonNull Application application) {
        super(application);
        mItemRepository = new ItemRepository(application);
        mFrozenItems = mItemRepository.getAllActiveFrozenItems();
    }


    public LiveData<List<FrozenItem>> getFrozenItems() {
        return mFrozenItems;
    }

    public void updateItem(FrozenItem item) {
        mItemRepository.updateItem(item);
    }

    public void delete(FrozenItem item) {
        mItemRepository.deleteItem(item);
    }

    public void archive(FrozenItem item) {
        mItemRepository.archiveItem(item);
    }

    public void restore(FrozenItem item) {
        mItemRepository.restoreItem(item);
    }

    public SortingOption.SortingOrder getSortingOrder() {
        return sortingOrder;
    }

    public void setSortingOrder(final SortingOption.SortingOrder sortingOrder) {
        this.sortingOrder = sortingOrder;
    }

    public SortingOption getSortingOption() {
        return sortingOption;
    }

    public void setSortingOption(final SortingOption sortingOption) {
        this.sortingOption = sortingOption;
    }

    /*
     * Notifications
     */

    /**
     * Get all scheduled notifications for the given item
     * @param item The item to retrieve the scheduled notifications for.
     * @return The list of notifications for the item.
     */
    public List<ItemNotification> getAllNotifications(FrozenItem item) {
        return mItemRepository.getAllNotifications(item);
    }

    /**
     * Cancels all notification workers for the given item.
     * @param itemToArchive The item for which the notification workers should be cancelled.
     */
    public void cancelNotifications(final FrozenItem itemToArchive) {
        ExecutorService executor = Executors.newSingleThreadExecutor();
        executor.execute(() -> {
            List<ItemNotification> allNotifications = getAllNotifications(itemToArchive);

            // Cancel all notifications
            WorkManager workManager = WorkManager.getInstance(getApplication());
            for (ItemNotification notification : allNotifications) {
                workManager.cancelWorkById(notification.getNotificationId());
                Log.d(TAG, "Cancelled the notification worker with uuid: " + notification.getNotificationId());
            }
        });
    }

    /**
     * Take the given amount from the given model and update the item in the repository
     * @param item The item to take from
     * @param amountTaken The amount that was taken from the item
     */
    public void takeFromItem(final FrozenItem item, final float amountTaken) {
        float newAmount = Math.max(0.0F, item.getAmount() - amountTaken);
        // TODO Check for possible side effects and possibly create copy of item first
        item.setAmount(newAmount);
        updateItem(item);
    }
}

File: app/src/main/java/com/mmutert/freshfreezer/viewmodel/ItemListViewModel.kt
package com.mmutert.freshfreezer.viewmodel

import android.app.Application
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.viewModelScope
import androidx.preference.PreferenceManager
import androidx.work.WorkManager
import com.mmutert.freshfreezer.data.Condition
import com.mmutert.freshfreezer.data.FrozenItem
import com.mmutert.freshfreezer.data.ItemDatabase.Companion.getDatabase
import com.mmutert.freshfreezer.data.ItemNotification
import com.mmutert.freshfreezer.data.ItemRepository
import com.mmutert.freshfreezer.util.SortingOption
import com.mmutert.freshfreezer.util.SortingOption.SortingOrder
import kotlinx.coroutines.launch
import java.util.concurrent.Executors
import kotlin.math.max

class ItemListViewModel(application: Application) : AndroidViewModel(application) {

    private val mItemRepository: ItemRepository = ItemRepository(getDatabase(application).itemDao())

    var frozenItems: LiveData<List<FrozenItem>> = mItemRepository.allActiveFrozenItems
        private set

    var sortingOrder = loadSortingOrderPreference()
        set(value) {
            field = value
            storeSortingOrderPreference()
        }
    var sortingOption = loadSortingOptionPreference()
        set(value) {
            field = value
            storeSortingOptionPreference()
        }

    fun filterItems(conditions: Collection<Condition?>) {
        frozenItems = Transformations.map(frozenItems) { input: List<FrozenItem> ->
            input.filter { conditions.contains(it.condition) }
        }
    }

    fun resetFilter() {
        frozenItems = mItemRepository.allActiveFrozenItems
    }

    private fun loadSortingOrderPreference(): SortingOrder {
        val preferences = PreferenceManager.getDefaultSharedPreferences(getApplication())
        val orderString = preferences.getString(SORTING_ORDER_KEY, "ASCENDING")
        return SortingOrder.valueOf(orderString!!)
    }

    private fun storeSortingOrderPreference() {
        val preferences = PreferenceManager.getDefaultSharedPreferences(getApplication())
        preferences.edit().putString(SORTING_ORDER_KEY, sortingOrder.toString()).apply()
    }

    private fun loadSortingOptionPreference(): SortingOption {
        val preferences = PreferenceManager.getDefaultSharedPreferences(getApplication())
        val orderString = preferences.getString(SORTING_OPTION_KEY, "DATE_BEST_BEFORE")

        return SortingOption.valueOf(orderString!!)
    }

    private fun storeSortingOptionPreference() {
        val preferences = PreferenceManager.getDefaultSharedPreferences(getApplication())
        preferences.edit().putString(SORTING_OPTION_KEY, sortingOption.toString()).apply()
    }

    fun updateItem(item: FrozenItem) {
        viewModelScope.launch {
            mItemRepository.updateItem(item)
        }
    }

    fun delete(item: FrozenItem) {
        viewModelScope.launch {
            mItemRepository.deleteItem(item)
        }
    }

    fun archive(item: FrozenItem) {
        viewModelScope.launch {
            mItemRepository.archiveItem(item)
        }
    }

    fun restore(item: FrozenItem) {
        viewModelScope.launch {
            mItemRepository.restoreItem(item)
        }
    }


    /*
     * Notifications
     */
    /**
     * Get all scheduled notifications for the given item
     * @param item The item to retrieve the scheduled notifications for.
     * @return The list of notifications for the item.
     */
    fun getAllNotifications(item: FrozenItem): List<ItemNotification> {
        return mItemRepository.getAllNotifications(item)
    }

    /**
     * Cancels all notification workers for the given item.
     * @param itemToArchive The item for which the notification workers should be cancelled.
     */
    fun cancelNotifications(itemToArchive: FrozenItem) {
        val executor = Executors.newSingleThreadExecutor()
        executor.execute {
            val allNotifications = getAllNotifications(itemToArchive)

            // Cancel all notifications
            val workManager = WorkManager.getInstance(getApplication())
            for ((_, notificationId) in allNotifications) {
                workManager.cancelWorkById(notificationId!!)
                Log.d(TAG, "Cancelled the notification worker with uuid: $notificationId")
            }
        }
    }

    /**
     * Take the given amount from the given model and update the item in the repository
     * @param item The item to take from
     * @param amountTaken The amount that was taken from the item
     */
    fun takeFromItem(item: FrozenItem, amountTaken: Float) {
        val newAmount = max(0.0f, item.amount - amountTaken)
        // TODO Check for possible side effects and possibly create copy of item first
        item.amount = newAmount
        updateItem(item)
    }

    companion object {
        const val TAG = "ItemListViewModel"
        private const val SORTING_ORDER_KEY = "sortingOrder"
        private const val SORTING_OPTION_KEY = "sortingOption"
    }
}